src/osd/OSD.cc

```diff
@@ -10536,6 +10536,7 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
       wait_lock.unlock();
     } else if (!sdata->stop_waiting) {
       dout(20) << __func__ << " empty q, waiting" << dendl;
+      sdata->shard_in_progress = false;
       osd->cct->get_heartbeat_map()->clear_timeout(hb);
       sdata->shard_lock.unlock();
       sdata->sdata_cond.wait(wait_lock);
```

Comment:
@tchaikov . move sdata->shard_in_progress = true  to this can fix the bug. For performance, i'll test.

Reply:
@majianpeng ping?